### PR TITLE
Add benchmarks for bloop

### PR DESCRIPTION
**Pull request by Duhemm**

### CODE REVIEW

File: benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
+package bloop
```
Comment (jvican) [inline benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala]: Good that we have them in our own project.

File: benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  var tempDir: Path = _
+  var processOutputReader: BufferedReader = _
+  var processInputReader: BufferedWriter = _
+  var output = new java.lang.StringBuilder()
+
+  @Setup(Level.Trial) def spawn(): Unit = {
+    tempDir = Files.createTempDirectory("bloop-")
+    val configDir = Files.createDirectory(tempDir.resolve(".bloop-config"))
+
+    val path = Paths.get(s"../frontend/src/test/resources/projects/$project")
```
Comment (jvican) [inline benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala]: What happens if the project doesn't exist? Shouldn't we report it?

File: build.sbt [hunk 3]
```diff
@@ -32,6 +32,13 @@ val nailgun = project
     hijackScalafmtOnCompile in SbtConfig in NailgunBuild := false,
   )
 
+val compilerBenchmark = project
```
Comment (jvican) [inline build.sbt]: I propose we call this benchmark bridge.

File: build.sbt [hunk 4]
```diff
@@ -82,6 +90,11 @@ val frontend = project
     parallelExecution in Test := false
   )
 
+val benchmarks = project
```
Comment (jvican) [inline build.sbt]: Do you add this to `allProjects`?

File: frontend/src/test/scala/bloop/tasks/IntegrationTestSuite.scala [hunk 5]
```diff
@@ -25,25 +25,9 @@ object IntegrationTestSuite extends DynTest {
...
-        javacOptions = Array.empty,
-        sourceDirectories = Array.empty,
-        previousResult = CompilationHelpers.emptyPreviousResult,
-        tmp = classesDir,
-        testFrameworks = Array.empty,
-        origin = None
-      )
-      projects + (rootProjectName -> rootProject)
-    }
+    val projects = ProjectHelpers.loadTestProject(testDirectory.getFileName.toString,
```
Comment (jvican) [inline frontend/src/test/scala/bloop/tasks/IntegrationTestSuite.scala]: Is it possible to avoid doing these changes? I change these things quite a bit, and looks unrelated to the PR?

File: project/BuildPlugin.scala [hunk 6]
```diff
@@ -68,6 +72,35 @@ object BuildKeys {
     Keys.libraryDependencies += Dependencies.utest,
     Keys.testFrameworks += new sbt.TestFramework("utest.runner.Framework"),
   )
+
+  import sbtassembly.AssemblyKeys
```
Comment (jvican) [inline project/BuildPlugin.scala]: :heart:

File: project/BuildPlugin.scala [hunk 7]
```diff
@@ -68,6 +72,35 @@ object BuildKeys {
...
+  val assemblySettings: Seq[Def.Setting[_]] = List(
+    Keys.mainClass in AssemblyKeys.assembly := Some("bloop.Bloop"),
+    Keys.test in AssemblyKeys.assembly := {}
+  )
+
+  val benchmarksSettings: Seq[Def.Setting[_]] = List(
+    Keys.skip in Keys.publish := true,
+    Keys.javaOptions ++= {
+      def refOf(version: String) = {
```
Comment (jvican) [inline project/BuildPlugin.scala]: You could also do it via dynver, which provides a very similar thing to this.

File: project/BuildPlugin.scala [hunk 8]
```diff
@@ -68,6 +72,35 @@ object BuildKeys {
...
+        }
+      }
+      List(
+        "-DscalaVersion=" + Keys.scalaVersion.value,
+        "-DscalaRef=" + refOf(Keys.scalaVersion.value),
+        "-Dsbt.launcher=" + (sys
+          .props("java.class.path")
+          .split(java.io.File.pathSeparatorChar)
+          .find(_.contains("sbt-launch"))
+          .getOrElse("")),
```
Comment (jvican) [inline project/BuildPlugin.scala]: Are we sure we want the empty string if sbt launch is not there? If this is a corner case, wouldn't it make more sense to throw an exception right away?

File: project/BuildPlugin.scala [hunk 9]
```diff
@@ -68,6 +72,35 @@ object BuildKeys {
     Keys.libraryDependencies += Dependencies.utest,
     Keys.testFrameworks += new sbt.TestFramework("utest.runner.Framework"),
   )
+
+  import sbtassembly.AssemblyKeys
+  val assemblySettings: Seq[Def.Setting[_]] = List(
```
Comment (jvican) [inline project/BuildPlugin.scala]: Will assembly directly add the nailgun server and all the bridges to this mega jar?

File: benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala (hunk 2 above)
Comment (Duhemm) [reply benchmarks/src/main/scala/bloop/HotBloopBenchmark.scala]: It'd crash a little after that, but let's throw an exception here.

File: project/BuildPlugin.scala (hunk 9 above)
Comment (Duhemm) [reply project/BuildPlugin.scala]: Yes, the nailgun server is part of the assembly, but not the bridges.

File: project/BuildPlugin.scala (hunk 8 above)
Comment (Duhemm) [reply project/BuildPlugin.scala]: You only need that when you want to benchmark sbt, so there are situations where it may be okay.

File: build.sbt (hunk 3 above)
Comment (Duhemm) [reply build.sbt]: I've changed it

**Comment by Duhemm** [comment]
I think I've addressed all the comments 🎉
